### PR TITLE
CUMULUS-2380: process dead letter archive

### PR DESCRIPTION
**Pull request by nemreid**
**Summary:** Summary of changes

Addresses [CUMULUS-2380: Handle messages that failed database write](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2380)

## Changes

- **CUMULUS-2380**
    - Add `process-dead-letter-archive` lambda to pick up and process dead letters in the S3 system bucket dead letter archive.
    - Add `/deadLetters` endpoint to trigger an async operation to leverage this capability on demand.
    - Add unit tests and integration test for all of the above.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [x] Integration tests

### CODE REVIEW

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+const { writeRecords } = require('./sf-event-sqs-to-db-records');
+
+async function processDeadLetterArchive({
+  knex,
+  bucket = process.env.system_bucket,
+  path = `${process.env.stackName}/dead-letter-archive/sqs/`,
+  writeRecordsFunction = writeRecords,
+}) {
+  // TODO: figure out a way to cap this so we don't read a billion objects into memory
```
Comment (nemreid) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:17]: Happy to take suggestions on this.

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  const deadLetterObjects = await listS3ObjectsV2({
+    Bucket: bucket,
+    Prefix: path,
+  });
```
Comment (nemreid) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:21]: Happy to take suggestions here.

Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js]: Resolved in https://github.com/nasa/cumulus/pull/2181/commits/54544cb48d8937dae00a79301bd382354964a4f0

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      testRule = await createOneTimeRule(
+        stackName,
+        {
+          workflow: 'HelloWorldWorkflow',
+          collection: pick(testCollection, ['name', 'version']),
+          provider: testProvider.id,
+          payload: { testId },
+        }
+      );
+      console.log('createOneTimeRule:', testRule);
```
Comment (botanical) [inline example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js:92]: Was this intentionally left in?

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  /* eslint-enable no-await-in-loop */
+}
+
+/**
+ * Lambda handler for AsyncOperation purposes
+ *
+ * @param {Object} event - Input payload object
+ * @param {string} [event.bucket] - Bucket containing dead letter archive (default to system bucket)
+ * @param {string} [event.key] - Dead letter archive path key
+ * @returns {Promise<undefined>}
```
Comment (botanical) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:63]: is there a difference between `Promise<undefined>` and `Promised<void>`? 🤔

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        granuleId: `MOD09GQ.${randomString()}.hdf`,
+        collectionId: constructCollectionId(testCollection.name, testCollection.version),
+        files: testFiles,
+      });
+
+      testRule = await createOneTimeRule(
+        stackName,
+        {
+          workflow: 'HelloWorldWorkflow',
+          collection: pick(testCollection, ['name', 'version']),
```
Comment (botanical) [inline example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js:87]: ✨ pick ✨   - how fancy

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js (hunk 3 above)
Comment (nemreid) [reply example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js:92]: No!

File: packages/api/lambdas/process-s3-dead-letter-archive.js (hunk 4 above)
Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js:63]: I guess `<void>` is the TS version, and `<undefined>` is the JS/ES version?

Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js:63]: It goes a [bit deeper than that](https://stackoverflow.com/questions/58885485/why-does-typescript-have-both-void-and-undefined) within TS, but I'm pretty sure `Promise<void>` isn't really appropriate in an EcmaScript file, as `void` there appears to be an operator with a distinct meaning.

Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js:63]: Then again, [ESLint](https://eslint.org/docs/rules/valid-jsdoc) appears to have the opinion that `@returns {void} constitutes valid JSDoc, so 🤷. We haven't done it anywhere else in JS files, but I don't mind starting now.

File: tf-modules/archive/process_dead_letter_archive.tf [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  environment {
+    variables = {
+      system_bucket   = var.system_bucket
+      stackName       = var.prefix
+      ExecutionsTable = var.dynamo_tables.executions.name
+      GranulesTable   = var.dynamo_tables.granules.name
+      PdrsTable       = var.dynamo_tables.pdrs.name
+      databaseCredentialSecretArn = var.rds_user_access_secret_arn
+      RDS_DEPLOYMENT_CUMULUS_VERSION = "5.0.0"
```
Comment (Jkovarik) [inline tf-modules/archive/process_dead_letter_archive.tf:100]: Should this lambda also take the RDS heartbeat env var?    For that matter, `write_db_dlq_records_to_s3` and even sf_event_sqs_to_db_records?       🤔

File: tf-modules/archive/api.tf [hunk 7]
```diff
@@ -35,6 +35,7 @@ locals {
       cmr_username                     = var.cmr_username
       databaseCredentialSecretArn      = var.rds_user_access_secret_arn
       dbHeartBeat                      = var.rds_connection_heartbeat
+      DeadLetterProcessingLambda       = aws_lambda_function.process_dead_letter_archive.arn
```
Comment (Jkovarik) [inline tf-modules/archive/api.tf:38]: Nit.... and not really your PR.  It'd be nice if we had some consistency in the env var casing.

File: packages/api-client/src/deadLetters.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * @param {Object} params              - params
+ * @param {string} params.prefix       - the prefix configured for the stack
+ * @param {Object} params.payload      - payload to post to the endpoint
+ * @param {Function} params.callback   - async function to invoke the api lambda
+ *                                       that takes a prefix / user payload.  Defaults
+ *                                       to cumulusApiClient.invokeApi
+ * @returns {Promise<Object>}          - promise that resolves to the output
+ *                                       of the API lambda
+ */
+export const postDeadLetters = async (params: {
```
Comment (Jkovarik) [inline packages/api-client/src/deadLetters.ts:17]: Technically this should have a unit.   We could certainly have a discussion about the unit tests in this package....

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+const log = require('@cumulus/common/log');
+
+const { s3 } = require('@cumulus/aws-client/services');
+const { getJsonS3Object } = require('@cumulus/aws-client/S3');
+const { getKnexClient } = require('@cumulus/db');
+
+const { writeRecords } = require('./sf-event-sqs-to-db-records');
+
+async function processDeadLetterArchive({
```
Comment (Jkovarik) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:11]: Particularly given this is .js : Docstring please!

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
+'use strict';
+
+const log = require('@cumulus/common/log');
+
+const { s3 } = require('@cumulus/aws-client/services');
+const { getJsonS3Object } = require('@cumulus/aws-client/S3');
+const { getKnexClient } = require('@cumulus/db');
+
+const { writeRecords } = require('./sf-event-sqs-to-db-records');
```
Comment (Jkovarik) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:9]: 👍🏻

File: packages/api/lambdas/process-s3-dead-letter-archive.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+async function processDeadLetterArchive({
+  knex,
+  bucket = process.env.system_bucket,
+  path = `${process.env.stackName}/dead-letter-archive/sqs/`,
+  writeRecordsFunction = writeRecords,
+  batchSize = 1000,
```
Comment (Jkovarik) [inline packages/api/lambdas/process-s3-dead-letter-archive.js:16]: Should this be configurable?

File: CHANGELOG.md [hunk 12]
```diff
@@ -134,9 +134,13 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
     - Changed data-migration1 and data-migration2 logic to only update already migrated records if the incoming record update has a newer timestamp
   - **CUMULUS-2329**
-    - Add write-db-dlq-records-to-s3 lambda.
+    - Add `write-db-dlq-records-to-s3` lambda.
     - Add terraform config to automatically write db records DLQ messages to an s3 archive on the system bucket.
     - Add unit tests and a component spec test for the above.
+  - **CUMULUS-2380**
+    - Add `process-dead-letter-archive` lambda to pick up and process dead letters in the S3 system bucket dead letter archive.
+    - Add `/deadLetters` endpoint to trigger an async operation to leverage this capability on demand.
+    - Add unit tests and integration test for all of the above.
```
Comment (Jkovarik) [inline CHANGELOG.md:143]: Do we have a ticket to document this feature?     I noticed this and the prior PR don't have any doc updates.

**Review by Jkovarik — COMMENTED** [review]
Most of the way there, dropping some comments while waiting for build/deploy to test.

File: packages/api/tests/lambdas/test-process-s3-dead-letter-archive.js [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    writeRecordsFunction: writeRecordsFunctionSpy,
+    batchSize: 15,
+  });
+  t.is(writeRecordsFunctionSpy.callCount, numberOfDeadLetters);
+  const remainingDeadLetters = await S3.listS3ObjectsV2({ Bucket: bucket, Prefix: path });
+  t.is(remainingDeadLetters.length, 0);
+});
+
+test('processDeadLetterArchive does not delete dead letters that fail to process', async (t) => {
+  const { bucket, path } = t.context;
```
Comment (Jkovarik) [inline packages/api/tests/lambdas/test-process-s3-dead-letter-archive.js:92]: 🐜  consider making this test have at least one success, or a logspy, or...   

Really I guess this feels like it should be more like 'processDeadLetterArchive only deletes dead letters that fail to process'.

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (error) {
+      beforeAllFailed = true;
+      console.log('beforeAll() failed, error:', error);
+    }
+  });
+
+  afterAll(async () => {
+    const ruleName = get(testRule, 'name');
+    if (ruleName) await deleteRule({ prefix: stackName, ruleName });
+    if (testCollection) {
```
Comment (Jkovarik) [inline example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js:141]: Shouldn't this fail if we wrote a granule?

File: packages/api/lambdas/process-s3-dead-letter-archive.js (hunk 11 above)
Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js:16]: Mostly there for unit test speed purposes, takes several minutes to spool through 1000-sized S3 pages and run the multiple-batch unit test.

File: CHANGELOG.md (hunk 12 above)
Comment (nemreid) [reply CHANGELOG.md:143]: Prior PR is automatic, but good point that this probably needs an operator doc.

File: packages/api/tests/lambdas/test-process-s3-dead-letter-archive.js (hunk 13 above)
Comment (nemreid) [reply packages/api/tests/lambdas/test-process-s3-dead-letter-archive.js:92]: 🤔 I don't think you wrote what you meant to write. The **intent** is _not_ to delete dead letters that fail to process. Only successfully processed dead letters are no longer necessary and safe to delete, the others must remain for manual inspection.

File: packages/api-client/src/deadLetters.ts (hunk 8 above)
Comment (nemreid) [reply packages/api-client/src/deadLetters.ts:17]: Fair point. I'm also interested in a naming debate on this endpoint.

File: packages/api/app/routes.js [hunk 15]
```diff
@@ -36,6 +37,9 @@ if (process.env.FAKE_AUTH === 'true') {
   ensureAuthorized = token.ensureAuthorized;
 }
 
+// dead letters endpoint
+router.use('/deadLetters', ensureAuthorized, deadLetters);
```
Comment (nemreid) [inline packages/api/app/routes.js:41]: Thoughts on the endpoint naming?

File: packages/api/lib/testUtils.js [hunk 16]
```diff
@@ -284,6 +285,28 @@ function fakeAccessTokenFactory(params = {}) {
   };
 }
 
+function fakeCumulusMessageFactory(params = {}) {
```
[the remark this answers is not in the document]
Comment (nemreid) [reply packages/api/lib/testUtils.js:288]: Oh man. I did this for DRY purposes so I'm gonna die on the hill of needing it here vs having to keep track of all the different places we already do this in tests. 😄 Talk about drift...

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js (hunk 14 above)
Comment (nemreid) [reply example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js:141]: 🤔 Probably.. I guess I should add a granule delete here too.

File: tf-modules/archive/process_dead_letter_archive.tf (hunk 6 above)
Comment (nemreid) [reply tf-modules/archive/process_dead_letter_archive.tf:100]: I'm not sure!

File: tf-modules/archive/api.tf (hunk 7 above)
Comment (nemreid) [reply tf-modules/archive/api.tf:38]: ..I don't disagree, but out of scope.. 😅

File: packages/api-client/src/deadLetters.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import { DeadLetterArchivePayload } from '@cumulus/types/api/dead_letters';
+import { invokeApi } from './cumulusApiClient';
+import { ApiGatewayLambdaHttpProxyResponse, InvokeApiFunction } from './types';
+
+/**
+ * POST a request to start a migrationCounts run
```
[the remark this answers is not in the document]
Comment (nemreid) [reply packages/api-client/src/deadLetters.ts:6]: 🤫

File: packages/api/app/routes.js (hunk 15 above)
Comment (Jkovarik) [reply packages/api/app/routes.js:41]: "There are two hard problems in computer science..." 

Really what this endpoint is doing re-running the database update for a cumulus message.  

/messageWrite ?  /messageRecover?

File: packages/api/tests/lambdas/test-process-s3-dead-letter-archive.js (hunk 13 above)
Comment (Jkovarik) [reply packages/api/tests/lambdas/test-process-s3-dead-letter-archive.js:92]: Indeed, I updated the comment, but the concern insofar as any nit does, is still there.

File: tf-modules/archive/process_dead_letter_archive.tf (hunk 6 above)
Comment (Jkovarik) [reply tf-modules/archive/process_dead_letter_archive.tf:100]: I think it should - we're using the heartbeat config 'globally' in the cumulus model - it causes the db model to run a verification query on a retry, which reduces performance but increases reliability in serverless situations where the db *might* be down.

File: packages/api/lib/testUtils.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/testUtils.js:288]: I'd suggest a ticket...but...

**Review by Jkovarik — CHANGES_REQUESTED** [review]
This all looks really good, I have a couple of comments I'd like to talk about/see resolved, and I'm thus far unable to demonstrate the integration test works on my CI or dev stack - we should chat tomorrow and see if we can get that resolved.

File: packages/api/app/routes.js (hunk 15 above)
Comment (npauzenga) [reply packages/api/app/routes.js:41]: Maybe something with a subpath like `/cumulusMessage/recover`?

Comment (markdboyd) [reply packages/api/app/routes.js:41]: I would suggest a subpath of `workflowMessages`. So `/deadLetters/workflowMessages`. It seems pretty clear then what type of data the endpoint is handling. You could go `/deadLetters/workflowMessages/process`, but not sure if it's not necessary

Comment (nemreid) [reply packages/api/app/routes.js:41]: Maybe `/deadLetterArchive/recoverCumulusMessages`

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js (hunk 14 above)
Comment (nemreid) [reply example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js:142]: https://github.com/nasa/cumulus/pull/2181/commits/89a057d0ac961b8e73adb5b153b9beed1da9759e

File: packages/api/lambdas/process-s3-dead-letter-archive.js (hunk 9 above)
Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js:23]: 7afb74661016d2cdd28a82ac61d8b174107a41ad

File: example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js (hunk 3 above)
Comment (nemreid) [reply example/spec/parallel/dbRecords/DeadLetterArchiveProcessingSpec.js]: https://github.com/nasa/cumulus/pull/2181/commits/2682e4efa888a2f54fd8fbd254d924c6e7bd3c7c

File: packages/api-client/src/deadLetters.ts (hunk 8 above)
Comment (nemreid) [reply packages/api-client/src/deadLetters.ts]: https://github.com/nasa/cumulus/pull/2181/commits/4b6a66a596cc05c31fd86bfa9f52cb254dd6ab14

File: packages/api-client/src/deadLetters.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (nemreid) [reply packages/api-client/src/deadLetters.ts]: https://github.com/nasa/cumulus/pull/2181/commits/4b6a66a596cc05c31fd86bfa9f52cb254dd6ab14

File: packages/api/app/routes.js (hunk 15 above)
Comment (nemreid) [reply packages/api/app/routes.js]: https://github.com/nasa/cumulus/pull/2181/commits/9f8fdf63e3c3cc0866389dae6ecfbb25a7ff00d2

File: packages/api/lambdas/process-s3-dead-letter-archive.js (hunk 4 above)
Comment (nemreid) [reply packages/api/lambdas/process-s3-dead-letter-archive.js]: 7afb74661016d2cdd28a82ac61d8b174107a41ad

File: tf-modules/archive/process_dead_letter_archive.tf (hunk 6 above)
Comment (nemreid) [reply tf-modules/archive/process_dead_letter_archive.tf:101]: https://github.com/nasa/cumulus/pull/2181/commits/1a1b6190970b18545ddac5359fb8bbb192c664a1

Comment (nemreid) [reply tf-modules/archive/process_dead_letter_archive.tf:101]: For the record, `write_db_dlq_records_to_s3` does not interact with the DB, only the SQS DLQ and S3.

**Review by Jkovarik — APPROVED** [review]
Thanks for addressing my concerns, nice work!

**Pull request merged**